File: services/api/backend/apps/api/routers/profile_stats.py
from __future__ import annotations

import uuid
from datetime import date
from typing import Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from backend.domain.gamification.stats_service import StatsService
from backend.infrastructure.database.session import get_session
from backend.infrastructure.security.auth import get_current_user_id

router = APIRouter(prefix="/api/v1/profile", tags=["profile"])


# ── Schemas ──


class BestDay(BaseModel):
    date: date
    seconds: int
    session_count: int


class BestWeek(BaseModel):
    week_start: date
    total_seconds: int
    per_day: list[int]


class BestMonth(BaseModel):
    month: str
    total_seconds: int
    heatmap: list[list[int]]


class ProfileStats(BaseModel):
    all_focus_seconds: int
    days_in_app: int
    longest_streak: int
    best_day: Optional[BestDay]
    best_week: Optional[BestWeek]
    best_month: Optional[BestMonth]


# ── Endpoint ──


@router.get("/stats", response_model=ProfileStats)
async def get_profile_stats(
    session: AsyncSession = Depends(get_session),
    current_user_id: uuid.UUID = Depends(get_current_user_id),
) -> ProfileStats:
    service = StatsService(session, current_user_id)
    data = await service.get_profile_stats()
    return ProfileStats(
        all_focus_seconds=data["all_focus_seconds"],
        days_in_app=data["days_in_app"],
        longest_streak=data["longest_streak"],
        best_day=BestDay(**data["best_day"]) if data["best_day"] else None,
        best_week=BestWeek(**data["best_week"]) if data["best_week"] else None,
        best_month=BestMonth(**data["best_month"]) if data["best_month"] else None,
    )

File: apps/mobile/lib/features/profile/models/profile_stats.dart
class BestDay {
  final DateTime date;
  final int seconds;
  final int sessionCount;

  const BestDay({
    required this.date,
    required this.seconds,
    required this.sessionCount,
  });

  factory BestDay.fromJson(Map<String, dynamic> json) {
    return BestDay(
      date: DateTime.parse(json['date'] as String),
      seconds: json['seconds'] as int? ?? 0,
      sessionCount: json['session_count'] as int? ?? 0,
    );
  }
}

class BestWeek {
  final DateTime weekStart;
  final int totalSeconds;
  final List<int> perDay;

  const BestWeek({
    required this.weekStart,
    required this.totalSeconds,
    required this.perDay,
  });

  factory BestWeek.fromJson(Map<String, dynamic> json) {
    final rawPerDay = json['per_day'];
    final perDay = rawPerDay is List
        ? rawPerDay.map((e) => e as int? ?? 0).toList()
        : <int>[];
    return BestWeek(
      weekStart: DateTime.parse(json['week_start'] as String),
      totalSeconds: json['total_seconds'] as int? ?? 0,
      perDay: perDay,
    );
  }
}

class BestMonth {
  final String month;
  final int totalSeconds;
  final List<List<int>> heatmap;

  const BestMonth({
    required this.month,
    required this.totalSeconds,
    required this.heatmap,
  });

  factory BestMonth.fromJson(Map<String, dynamic> json) {
    final rawHeatmap = json['heatmap'];
    List<List<int>> heatmap = [];
    if (rawHeatmap is List) {
      heatmap = rawHeatmap.map((row) {
        if (row is List) {
          return row.map((e) => e as int? ?? 0).toList();
        }
        return <int>[];
      }).toList();
      // Tolerate up to 6 rows — keep as-is; the UI decides how to render
    }
    return BestMonth(
      month: json['month'] as String? ?? '',
      totalSeconds: json['total_seconds'] as int? ?? 0,
      heatmap: heatmap,
    );
  }
}

class ProfileStats {
  final int allFocusSeconds;
  final int daysInApp;
  final int longestStreak;
  final BestDay? bestDay;
  final BestWeek? bestWeek;
  final BestMonth? bestMonth;

  const ProfileStats({
    required this.allFocusSeconds,
    required this.daysInApp,
    required this.longestStreak,
    this.bestDay,
    this.bestWeek,
    this.bestMonth,
  });

  factory ProfileStats.fromJson(Map<String, dynamic> json) {
    final rawBestDay = json['best_day'];
    final rawBestWeek = json['best_week'];
    final rawBestMonth = json['best_month'];

    return ProfileStats(
      allFocusSeconds: json['all_focus_seconds'] as int? ?? 0,
      daysInApp: json['days_in_app'] as int? ?? 0,
      longestStreak: json['longest_streak'] as int? ?? 0,
      bestDay: rawBestDay is Map<String, dynamic>
          ? BestDay.fromJson(rawBestDay)
          : null,
      bestWeek: rawBestWeek is Map<String, dynamic>
          ? BestWeek.fromJson(rawBestWeek)
          : null,
      bestMonth: rawBestMonth is Map<String, dynamic>
          ? BestMonth.fromJson(rawBestMonth)
          : null,
    );
  }
}
